pkg/sql/plpgsql/parser/parse.go

```diff
@@ -0,0 +1,171 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package parser
+
+import (
+	"go/constant"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/scanner"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/plpgsqltree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/errors"
+)
+
+func init() {
+	scanner.NewNumValFn = func(a constant.Value, s string, b bool) interface{} { return tree.NewNumVal(a, s, b) }
+	scanner.NewPlaceholderFn = func(s string) (interface{}, error) { return tree.NewPlaceholder(s) }
+}
+
+// Statement is the result of parsing a single statement. It contains the AST
+// node along with other information.
+type Statement struct {
+	// AST is the root of the AST tree for the parsed statement.
+	AST *plpgsqltree.PLpgSQLStmtBlock
+
+	// SQL is the original SQL from which the statement was parsed. Note that this
+	// is not appropriate for use in logging, as it may contain passwords and
+	// other sensitive data.
+	SQL string
+
+	// NumPlaceholders indicates the number of arguments to the statement (which
+	// are referenced through placeholders). This corresponds to the highest
+	// argument position (i.e. the x in "$x") that appears in the query.
+	//
+	// Note: where there are "gaps" in the placeholder positions, this number is
+	// based on the highest position encountered. For example, for `SELECT $3`,
+	// NumPlaceholders is 3. These cases are malformed and will result in a
+	// type-check error.
+	NumPlaceholders int
+
+	// NumAnnotations indicates the number of annotations in the tree. It is equal
+	// to the maximum annotation index.
+	NumAnnotations tree.AnnotationIdx
+}
+
+// String returns the AST formatted as a string.
+func (stmt Statement) String() string {
+	return stmt.StringWithFlags(tree.FmtSimple)
+}
+
+// StringWithFlags returns the AST formatted as a string (with the given flags).
+func (stmt Statement) StringWithFlags(flags tree.FmtFlags) string {
+	ctx := tree.NewFmtCtx(flags)
+	stmt.AST.Format(ctx)
+	return ctx.CloseAndGetString()
+}
+
+// Parser wraps a scanner, parser and other utilities present in the parser
+// package.
+type Parser struct {
+	scanner    scanner.PLPGSQLScanner
+	lexer      lexer
+	parserImpl plpgsqlParserImpl
+	tokBuf     [8]plpgsqlSymType
+}
+
+// INT8 is the historical interpretation of INT. This should be left
+// alone in the future, since there are many sql fragments stored
+// in various descriptors. Any user input that was created after
+// INT := INT4 will simply use INT4 in any resulting code.
+var defaultNakedIntType = types.Int
+
+// Parse parses the sql and returns a list of statements.
+func (p *Parser) Parse(sql string) (Statement, error) {
+	return p.parseWithDepth(1, sql, defaultNakedIntType)
+}
+
+func (p *Parser) scanOneStmt() (sql string, tokens []plpgsqlSymType, done bool) {
```

Comment:
We'd want to rename this to something like `scanFnBlock` since it's not scanning one statement, but the whole function definition. I'm bad at naming things, so your call :)